diff --git a/kernel/sys.c b/kernel/sys.c
index b276a5e..898dbf1 100644
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -1259,5 +1259,9 @@ static int override_release(char __user *release, size_t len)
 return ret;
 }
 
+#ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
+extern struct static_key_false susfs_is_uname_spoof_buffer_set;
+extern void susfs_spoof_uname(struct new_utsname* tmp);
+#endif
 static int override_version(struct new_utsname __user *name)
 {
 #ifdef CONFIG_F2FS_REPORT_FAKE_KERNEL_VERSION
@@ -1290,5 +1294,9 @@ SYSCALL_DEFINE1(newuname, struct new_utsname __user *, name)

	down_read(&uts_sem);
	memcpy(&tmp, utsname(), sizeof(tmp));
+#ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
+	if (static_branch_likely(&susfs_is_uname_spoof_buffer_set))
+		susfs_spoof_uname(&tmp);
+#endif	
	up_read(&uts_sem);

	rcu_read_lock();
	for_each_thread(current, t) {
